### ISSUE TITLE
multiple WebAppContext can have same context path

### ISSUE THREAD

**Issue by Artin-Salehi** [eclipse/jetty.project#9877]
hello,
jetty :12.0.0.beta1
WebAppContext module : jetty-ee10-webapp
Example :
		var appCol = new ContextHandlerCollection();
		var restCtx= new WebAppContext("webapps/rest", "/app");	
		var shopCtx= new WebAppContext("webapps/shop", "/app");
		
		appCol.addHandler(restCtx);
		appCol.addHandler(shopCtx);
		
		server.setHandler(appCol);

**Comment by sbordet** [comment]
That multiple `WebAppContext` can have the same context path is a feature, and it is heavily used, so that is completely normal.

Please make some effort to describe with more precision what you think the issue is.

**Comment by Artin-Salehi** [comment]
Thanks for your reply, but I have one more question.
If the WebAppContext represents a servlet-based web application, shouldn't it have its own context path by standard?
Even the Jetty documentation itself says that you can have multiple web applications with different paths and even add these applications to the server dynamically, assuming one of the applications has the same path as before. What should happen?
I haven't seen anything like this in Tomcat, I agree with you that it would be a nice feature but isn't it against the standard?! And can this behavior be trusted? Is it possible to deploy new web applications at runtime without realizing that the context path might be the same? Even if the path is specified in the default-context-path element in web.xml?
Or should the programmer do this and check the context path before deployment?
Thank you.

**Comment by joakime** [comment]
The context-path is often seen as the prefix in url path space for all requests.
But that's a limited view of it's scope.

The url path isn't the only thing determining if the request reaches that WebAppContext.
Things like SNI, DNS based Virtual Hosts,  InetAccessHandlers, Named connector Virtual Hosts, etc. (pretty much anything that a Handler before your WebAppContext can use to determine if the request reaches your WebAppContext or not)

**Comment by sbordet** [comment]
>  If the WebAppContext represents a servlet-based web application, shouldn't it have its own context path by standard?

There is no standard as to how to deploy a web application.

Commonly, Servlet Containers have standardized to use the name of the `*.war` file as the context path, but every Servlet Container offers more configuration options (specific to each Servlet Container) to deploy web applications.

I'm not sure what your concern is, but if you use the `*.war` filename as the context path, then obviously the `*.war` filename is the same, and that file represents one web application. If you overwrite it with a different web application that has the same name, then it's your mistake.

If you use `default-context-path` in `web.xml` it's typically not a good idea, and in any case it can be overridden by a specific configuration of the Servlet Container, which is what Jetty allows you to.

Without this configuration capabilities you would not be able to deploy 2 different websites to the same Servlet Container, for example `beautifulhome.com` and `nicevacations.com` -- both need to be deployed at context path `/` and the Servlet Container can differentiate them via the virtual host configuration that you must provide.

In summary, it's not against any standard, and this behavior can be trusted. Developers must be aware and the deployer of the web application must know to what domain and context path the web application should be deployed and how to configure the Servlet Container to achieve that.

**Comment by janbartel** [comment]
I think this question seems to have been answered. If you have further concerns please reopen with fresh info.
